Make `ArrayBuffer`'s iterator fail fast when buffer is mutated

NthPortal opened the pull request:
Make `ArrayBuffer`'s iterator fail-fast when the buffer is mutated after the iterator's creation.

Fix inserting an `ArrayBuffer` into itself (scala/bug#12121).

Partially addresses scala/bug#12121 and scala/bug#12009

NthPortal committed:
[bug#12009] Make ArrayBuffer's iterator fail-fast

Make `ArrayBuffer`'s iterator fail-fast when the buffer is
mutated after the iterator's creation.

NthPortal committed:
[bug#12121] Add test for inserting an ArrayBuffer into itself

NthPortal commented:
I wonder if these commits from a month ago will Just Work™

Ichoran commented on src/library/scala/collection/IndexedSeq.scala:
```diff
@@ -47,15 +47,7 @@ trait IndexedSeqOps[+A, +CC[_], +C] extends Any with SeqOps[A, CC, C] { self =>
-        self(i)
-      } else Iterator.empty.next()
-  }
+  override def reverseIterator: Iterator[A] = view.reverseIterator
```
Maybe it's not used enough to matter, but are you sure this isn't going to result in a sizable performance hit?  The view reverse iterator does quite a bit more work, and may require double-bounds-checks for indexing.

NthPortal replied:
the view reverse iterator doesn't do that much more work. also, the regular iterator is already implemented by `view.iterator`, so I don't see it being any worse than that

Ichoran commented:
You didn't by any chance try any benchmarks with this, did you?

NthPortal commented:
I don't have a good benchmarking setup, sorry

NthPortal commented:
(I will squash out that merge, dw)

SethTisue commented:
is there somewhere we've gathered together knowledge/advice on how to benchmark stuff in scala/scala?

I don't know because I also have left the benchmarking to others :-)

if no such single place exists, maybe at minimum we could start a scala-dev ticket as a place to collect knowledge

NthPortal commented:
My issue is that all I have is my laptop, with the following issues:

- it's a laptop, so it's not the most reliable in the first place, due to thermal throttling
- it tends to have lots of applications idle/running in the background
- I can't use it while benchmarks are running

dwijnand commented:
Needs a build.sbt rebase :(

NthPortal commented:
it is done

som-snytt commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -61,21 +64,27 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
+  protected def ensureSize(n: Int): Unit = {
+    mutationCount += 1
     array = ArrayBuffer.ensureSize(array, size0, n)
+  }
```
There's also https://github.com/scala/scala/pull/9275

dwijnand commented:
We moved the MiMa filters, so this needs a rebase again (sorry!)

NthPortal commented:
done again

dwijnand commented on src/library/scala/collection/mutable/CheckedIndexedSeqView.scala:
```diff
@@ -0,0 +1,117 @@
+  override def prependedAll[B >: A](prefix: IndexedSeqView.SomeIndexedSeqOps[B]): IndexedSeqView[B] = new CheckedIndexedSeqView.Concat(prefix, this)(mutationCount)
+}
+
+private object CheckedIndexedSeqView {
```
private object doesn't mean private? 🤯

NthPortal replied:
it means `private[mutable]`

lrytz replied:
https://www.scala-lang.org/files/archive/spec/2.13/09-top-level-definitions.html#packagings

> If a definition in ds is labeled private, it is visible only for other members in the package.

But to me `package p { private class C }` looks rather uncommon, it wasn't intuitively clear to me what it means. I'd make `private[p]` explicit.

som-snytt replied:
Recently, I learned about the mindblown emoji and also that `private` is not the same as `private[x]` where `x` is the enclosing element. 🤯 !

NthPortal replied:
i'll try to get to it soon. if i don't and someone wants to get this merged, feel free to push to my branch

NthPortal replied:
fixed

NthPortal commented:
Fixing `patchInPlace` for scala/bug#12121 is blocked by #9141, since it requires adding an override that returns `this.type`

SethTisue commented:
/rebuild

SethTisue commented:
@julienrf you don't think we need benchmarks before merging?

julienrf commented:
@SethTisue Good point, yes benchmarks would be welcome.

SethTisue commented:
Okay, I'll benchmark it in time for 2.13.5.

SethTisue commented:
looks like I volunteered myself for some benchmarking. doh!

thanks to the pandemic, I have a spare Lightbend laptop right here, getting it set up for that, in the not-really-that-rigorous way one does

also reading those benchmarking instructions I'm always telling other people they should read

SethTisue commented:
Now that the 2.13.5 timeline is shortened, I'm afraid I won't have time to deal with this until afterwards, sorry :-(

NthPortal commented:
ping @SethTisue

SethTisue commented:
I do still intend to get to this in time for 2.13.6.

lrytz commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -160,6 +160,8 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
   def insertAll(@deprecatedName("n", "2.13.0") index: Int, elems: IterableOnce[A]): Unit = {
     checkWithinBounds(index, index)
     elems match {
+      case elems: AnyRef if elems eq this =>
+        insertAll(index, ArrayBuffer.from(this))
```
This seems no longer necessary? 2.13.5 and current 2.13.x:

```
scala> val b = scala.collection.mutable.ArrayBuffer(1, 2, 3)
val b: scala.collection.mutable.ArrayBuffer[Int] = ArrayBuffer(1, 2, 3)

scala> b.insertAll(1, b)

scala> b
val res1: scala.collection.mutable.ArrayBuffer[Int] = ArrayBuffer(1, 1, 2, 3, 2, 3)
```

lrytz replied:
Here's a bug (?) though, create a ticket?

```
scala> val b = scala.collection.mutable.ArrayBuffer(1, 2, 3)
val b: scala.collection.mutable.ArrayBuffer[Int] = ArrayBuffer(1, 2, 3)

scala> b.insertAll(1, b.view)

scala> b
val res16: scala.collection.mutable.ArrayBuffer[Int] = ArrayBuffer(1, 1, 1, 1, 2, 3)
```

lrytz replied:
(I just saw https://github.com/scala/bug/issues/12284 / https://github.com/scala/scala/pull/9388. 2.13.5-bin-55f4b15-SNAPSHOT from #9388 shows the same behavior as above).

NthPortal replied:
we don't really currently have a model that solves the `.view`-based bug (for this or any other collection); I believe it would have to be addressed on a collection-by-collection basis. `ListBuffer` is probably the only collection that currently deliberately avoids the issue, as of #9257. I think the bug is related to scala/bug#12121, and Seb commented about this issue specifically https://github.com/scala/bug/issues/12121#issuecomment-675741476.

I have an experimental branch that is very much a work in progress where the `ArrayBuffer` knows that it's segmented/split during an `insertAll` call, and can access elements properly using `apply` despite that, and that would fix the bug for `ArrayBuffer` specifically. I also don't know what the performance impact of it would be. I'm generally in the camp that correctness trumps performance, but I know that not everyone agrees with me in all cases.

NthPortal replied:
I've been looking at the current implementation of `ArrayBuffer#insertAll` for several minutes now, and I'm struggling to understand why the first example works without this PR. I will continue to investigate, and perhaps remove this case when I can convince myself why it's not needed (and add a note explaining)

NthPortal replied:
I believe the reason `b.appendAll(1, b)` works currently is because after copying the end of the array to the right to make space, it doesn't clear the original values from where they were, and `System.arraycopy` "atomically" copies the array into itself.

NthPortal replied:
I will remove the extra case

lrytz commented:
LGTM, leaving for @SethTisue to test / benchmark / merge.

NthPortal commented:
I've taken care of rebasing

SethTisue commented:
Note that this is a sequel to #9174, where the same thing was done but for `ListBuffer` instead of `ArrayBuffer`.

So that gives us some hope that this is a reasonably safe change; it may break some code but likely not very much, and we'll get additional evidence for that from community build results.

And it also gives us some hope that the benchmark results won't surprise us.

SethTisue commented:
I have started working on this, but it's not going to make 2.13.6. I'll make it a priority for early in the 2.13.7 process.

NthPortal commented:
@SethTisue do you have any benchmarks you're already running? I finally have a machine to run benchmarks on

som-snytt commented:
I was at Costco yesterday and they already have Halloween costumes.

SethTisue committed:
Update and reorganize benchmarks readme to be less confusing

We have exactly one custom runner, so the emphasis that the old
structure put on custom runners was misleading. It's better to
foreground the normal case.

SethTisue commented:
> I'll make it a priority for early in the 2.13.7 process.

In order not to make a liar of myself, I am unilaterally delaying the release of 2.13.7 is delayed until 2024.

SethTisue commented:
> The readme under test/benchmarks is okay (not perfect, but serviceable)

I've pushed a round of improvements.

SethTisue committed:
Add ArrayBufferBenchmark

Copy-pasted from `ListBufferBenchmark`, but based on the specifics of
the code changes in this PR, I added benchmarks for `addAll` and
`reverseIterator`, and modified the `insertAll` benchmark so it's
measuring `ArrayBuffer`-into-`ArrayBuffer` insertion (rather than
`Seq`-into-`ArrayBuffer`).

SethTisue commented:
@NthPortal I've pushed a proposed `ArrayBufferBenchmark`. if you feel like running it yourself go ahead, otherwise I'll run it myself next time there is safely breathable air in Reno where my spare laptop is

som-snytt commented:
> I wonder if these commits from a year ago will Just Work™

NthPortal commented:
I need a sobbing emoji reaction :sob:

NthPortal commented:
@SethTisue I've got the latest commit benchmarking in a screen on my server. once that's done, I'll rebase and reorder the commits, and run the benchmark before my changes

SethTisue commented:
okay. I'm running the benchmarks now myself as well — did an "after" run overnight, kicking off a "before" run now. I intend to try to get some graphs out of https://github.com/lightbend/benchdb , and to add some instructions for that to the benchmarking readme

SethTisue commented:
results on JDK 11 are below. I'd summarize as:

* `addAll`: no difference
* `filterInPlace`: 5% speedup
* `flatMapInPlace1`: inconsistent results, but seems like a wash
* `iteratorA`: inconsistent results, but looks like 10% slower?
* `iteratorB`: ditto
* `reverseIteratorA`: 2x speedup on long enough input
* `reverseIteratorB`: bigger and bigger slowdown as input gets longer
* `update`: inconsistent results, but seems like a wash

I don't know what to make of this, especially as a first-time JMH-er. (The last time I did any really serious benchmarking and performance tuning work on the JVM was circa 2005.)

@NthPortal I see you got similarly difficult-to-interpret results when you benchmarked your `ListBuffer` change on #9174

@Ichoran any clue what to make of this?

Suppose we assume these numbers are actually accurate/meaningful, then the concerning entries are:

* `iteratorA`/`iteratorB`: does the slowdown seem unexpectedly large? perhaps we should do a further round of benchmarking on just these two and make sure we are getting consistent numbers?
* the `reverseIterator` benchmarks. These are pretty big differences. It seems we are making some different tradeoff now. @NthPortal do you understand what's happening here?

```diff
% paste -d '\n' =(lam -s- before.txt) =(lam -s+ after.txt)
-[info] Benchmark                              (size)  Mode  Cnt       Score     Error  Units
+[info] Benchmark                              (size)  Mode  Cnt       Score      Error  Units
-[info] ArrayBufferBenchmark.addAll                10  avgt   30      95.879 ±   0.327  ns/op
+[info] ArrayBufferBenchmark.addAll                10  avgt   30      96.068 ±    0.311  ns/op
-[info] ArrayBufferBenchmark.addAll               100  avgt   30     247.127 ±   2.074  ns/op
+[info] ArrayBufferBenchmark.addAll               100  avgt   30     247.423 ±    2.500  ns/op
-[info] ArrayBufferBenchmark.addAll              1000  avgt   30    1381.755 ±  29.246  ns/op
+[info] ArrayBufferBenchmark.addAll              1000  avgt   30    1385.090 ±   29.158  ns/op
-[info] ArrayBufferBenchmark.addAll             10000  avgt   30   17222.398 ± 234.876  ns/op
+[info] ArrayBufferBenchmark.addAll             10000  avgt   30   17318.745 ±  328.651  ns/op
-[info] ArrayBufferBenchmark.filterInPlace         10  avgt   30      74.448 ±   0.305  ns/op
+[info] ArrayBufferBenchmark.filterInPlace         10  avgt   30      70.480 ±    2.179  ns/op
-[info] ArrayBufferBenchmark.filterInPlace        100  avgt   30     482.087 ±   2.803  ns/op
+[info] ArrayBufferBenchmark.filterInPlace        100  avgt   30     459.823 ±   10.969  ns/op
-[info] ArrayBufferBenchmark.filterInPlace       1000  avgt   30    4298.946 ±  17.671  ns/op
+[info] ArrayBufferBenchmark.filterInPlace       1000  avgt   30    3523.871 ±   13.017  ns/op
-[info] ArrayBufferBenchmark.filterInPlace      10000  avgt   30   42661.635 ± 147.395  ns/op
+[info] ArrayBufferBenchmark.filterInPlace      10000  avgt   30   40934.593 ±  197.500  ns/op
-[info] ArrayBufferBenchmark.flatMapInPlace1       10  avgt   30     281.237 ±   2.356  ns/op
+[info] ArrayBufferBenchmark.flatMapInPlace1       10  avgt   30     289.048 ±    4.302  ns/op
-[info] ArrayBufferBenchmark.flatMapInPlace1      100  avgt   30    2265.730 ±   9.631  ns/op
+[info] ArrayBufferBenchmark.flatMapInPlace1      100  avgt   30    2822.452 ±  387.805  ns/op
-[info] ArrayBufferBenchmark.flatMapInPlace1     1000  avgt   30   20813.614 ±  99.126  ns/op
+[info] ArrayBufferBenchmark.flatMapInPlace1     1000  avgt   30   20204.314 ±  209.202  ns/op
-[info] ArrayBufferBenchmark.flatMapInPlace1    10000  avgt   30  207020.049 ± 624.000  ns/op
+[info] ArrayBufferBenchmark.flatMapInPlace1    10000  avgt   30  209867.348 ±  691.503  ns/op
-[info] ArrayBufferBenchmark.iteratorA             10  avgt   30      46.072 ±   0.148  ns/op
+[info] ArrayBufferBenchmark.iteratorA             10  avgt   30      51.087 ±    0.249  ns/op
-[info] ArrayBufferBenchmark.iteratorA            100  avgt   30     231.861 ±   7.854  ns/op
+[info] ArrayBufferBenchmark.iteratorA            100  avgt   30     221.700 ±    0.741  ns/op
-[info] ArrayBufferBenchmark.iteratorA           1000  avgt   30    1778.509 ±   6.242  ns/op
+[info] ArrayBufferBenchmark.iteratorA           1000  avgt   30    1787.296 ±  145.555  ns/op
-[info] ArrayBufferBenchmark.iteratorA          10000  avgt   30   18767.094 ±  99.698  ns/op
+[info] ArrayBufferBenchmark.iteratorA          10000  avgt   30   20552.689 ± 1959.058  ns/op
-[info] ArrayBufferBenchmark.iteratorB             10  avgt   30     110.024 ±   0.253  ns/op
+[info] ArrayBufferBenchmark.iteratorB             10  avgt   30     113.312 ±    0.292  ns/op
-[info] ArrayBufferBenchmark.iteratorB            100  avgt   30     593.697 ±   9.560  ns/op
+[info] ArrayBufferBenchmark.iteratorB            100  avgt   30     649.757 ±    4.604  ns/op
-[info] ArrayBufferBenchmark.iteratorB           1000  avgt   30    4737.180 ±  21.966  ns/op
+[info] ArrayBufferBenchmark.iteratorB           1000  avgt   30    5792.775 ±  137.734  ns/op
-[info] ArrayBufferBenchmark.iteratorB          10000  avgt   30   50537.713 ± 200.756  ns/op
+[info] ArrayBufferBenchmark.iteratorB          10000  avgt   30   58559.043 ±  180.187  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA      10  avgt   30      50.080 ±   0.179  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA      10  avgt   30      44.190 ±    0.325  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA     100  avgt   30     213.165 ±   0.811  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA     100  avgt   30     151.188 ±    0.603  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA    1000  avgt   30    2058.113 ±  10.705  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA    1000  avgt   30    1156.087 ±    6.653  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA   10000  avgt   30   21187.332 ±  91.830  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA   10000  avgt   30   12620.373 ±   94.847  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB      10  avgt   30      97.431 ±   0.243  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB      10  avgt   30     100.641 ±    0.383  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB     100  avgt   30     530.075 ±   5.120  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB     100  avgt   30     671.618 ±   23.903  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB    1000  avgt   30    4875.297 ± 143.256  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB    1000  avgt   30    5817.915 ±   24.379  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB   10000  avgt   30   42559.799 ± 516.623  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB   10000  avgt   30   61057.526 ± 1408.570  ns/op
-[info] ArrayBufferBenchmark.update                10  avgt   30      41.757 ±   0.200  ns/op
+[info] ArrayBufferBenchmark.update                10  avgt   30      40.358 ±    0.135  ns/op
-[info] ArrayBufferBenchmark.update               100  avgt   30     207.730 ±   1.015  ns/op
+[info] ArrayBufferBenchmark.update               100  avgt   30     196.894 ±    1.252  ns/op
-[info] ArrayBufferBenchmark.update              1000  avgt   30    1801.087 ±  10.438  ns/op
+[info] ArrayBufferBenchmark.update              1000  avgt   30    1799.125 ±    8.619  ns/op
-[info] ArrayBufferBenchmark.update             10000  avgt   30   17487.817 ±  86.834  ns/op
+[info] ArrayBufferBenchmark.update             10000  avgt   30   18531.185 ±   95.866  ns/op
```

(I'm still fooling with getting https://github.com/lightbend/benchdb to render this data in graph form.)

Ichoran commented:
@SethTisue - The speed changes don't look too terrible to me.  I do wonder about the differences between the iterator benchmarks.  I suspect that this is hinting that the code optimizes well in a simple context, but less so in a complex context, since `toVector` in `iteratorB` is less straightforward than the `+=` in `iteratorA`.

So I would recommend a yet more complicated test, and a rerun of `iteratorA` and `iteratorB` to make sure they're consistent.  I don't have a great idea right now of what more complicated test would be good.  Maybe something with `.buffered` so we get an additional layer of indirection.

NthPortal commented on src/library/scala/collection/IndexedSeqView.scala:
```diff
@@ -82,18 +83,18 @@ object IndexedSeqView {
+    @inline private[this] def _hasNext: Boolean = remainder > 0
+    def hasNext: Boolean = _hasNext
     def next(): A =
-      if (pos < 0) throw new NoSuchElementException
```
this was a bug/mistake, btw. it should have been checking `remainder`, not `pos`

NthPortal commented:
so, the change to `reverseIterator` is that we now just use the one from `IterableSeqView`, which has optimisations for `drop` and `slice` operations (and possibly implements `take` with `slice`?). since those methods aren't benchmarked, I'm not sure how the optimisations would end up helping in one of the cases. I could see the overhead of the optimisations hurting in the other case, but not by that much

SethTisue commented:
I re-ran just the iterator benchmarks, and the results came out pretty much the same as last time. So that's good. I think we can consider the `iteratorA`/`iteratorB` benchmarks resolved now. The `reverse` numbers remain puzzling -- in the same way as before.

@NthPortal but you've now pushed some code changes; these numbers are still for the old code. Would you expect your changes to affect the `reverse` numbers?

I'm leaving for Chicago tomorrow (Reno = 🔥😷) and it wouldn't be smart traveling with both of my laptops, so I won't able to run more benchmarks unless I get set up to run them on an EC2 node or something like that... which would ultimately be better, actually...

```diff
 [info] Benchmark                              (size)  Mode  Cnt      Score     Error  Units
-[info] ArrayBufferBenchmark.iteratorA             10  avgt   30     45.931 ±   0.098  ns/op
+[info] ArrayBufferBenchmark.iteratorA             10  avgt   30     51.451 ±   0.274  ns/op
-[info] ArrayBufferBenchmark.iteratorA            100  avgt   30    231.938 ±   8.610  ns/op
+[info] ArrayBufferBenchmark.iteratorA            100  avgt   30    222.086 ±   0.811  ns/op
-[info] ArrayBufferBenchmark.iteratorA           1000  avgt   30   1779.704 ±   6.991  ns/op
+[info] ArrayBufferBenchmark.iteratorA           1000  avgt   30   1589.157 ±   5.483  ns/op
-[info] ArrayBufferBenchmark.iteratorA          10000  avgt   30  18837.041 ±  94.648  ns/op
+[info] ArrayBufferBenchmark.iteratorA          10000  avgt   30  18302.873 ± 424.876  ns/op
-[info] ArrayBufferBenchmark.iteratorB             10  avgt   30    110.579 ±   0.561  ns/op
+[info] ArrayBufferBenchmark.iteratorB             10  avgt   30    113.672 ±   0.344  ns/op
-[info] ArrayBufferBenchmark.iteratorB            100  avgt   30    582.791 ±   3.183  ns/op
+[info] ArrayBufferBenchmark.iteratorB            100  avgt   30    634.847 ±   1.702  ns/op
-[info] ArrayBufferBenchmark.iteratorB           1000  avgt   30   4739.360 ±  15.559  ns/op
+[info] ArrayBufferBenchmark.iteratorB           1000  avgt   30   5706.898 ±  69.042  ns/op
-[info] ArrayBufferBenchmark.iteratorB          10000  avgt   30  50616.912 ± 226.024  ns/op
+[info] ArrayBufferBenchmark.iteratorB          10000  avgt   30  58735.730 ± 177.797  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA      10  avgt   30     48.998 ±   0.431  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA      10  avgt   30     44.500 ±   0.356  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA     100  avgt   30    240.564 ±  16.821  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA     100  avgt   30    153.563 ±   1.672  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA    1000  avgt   30   2172.753 ±  80.909  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA    1000  avgt   30   1164.767 ±   5.038  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorA   10000  avgt   30  19996.335 ± 798.085  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorA   10000  avgt   30  12692.784 ±  81.984  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB      10  avgt   30    100.182 ±   1.392  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB      10  avgt   30    100.638 ±   0.489  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB     100  avgt   30    589.141 ±   1.790  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB     100  avgt   30    703.322 ±   1.997  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB    1000  avgt   30   4676.357 ±  13.902  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB    1000  avgt   30   5806.503 ±  25.428  ns/op
-[info] ArrayBufferBenchmark.reverseIteratorB   10000  avgt   30  42091.599 ± 212.420  ns/op
+[info] ArrayBufferBenchmark.reverseIteratorB   10000  avgt   30  59663.847 ± 875.264  ns/op
```

NthPortal commented:
@SethTisue I just rearranged the commits (and did some squashing/message cleanup) so that the benchmarks came before the changes, to make benchmarking before the changes easier. I didn't actually make any meaningful changes. in retrospect, I could also have temporarily reverted the changes.

I also rebased, for cleanliness

NthPortal commented:
run on Java 16

```diff
 Benchmark                              (size)  Mode  Cnt       Score      Error  Units
-ArrayBufferBenchmark.addAll                10  avgt   30      97.711 ±    0.594  ns/op
+ArrayBufferBenchmark.addAll                10  avgt   30     100.307 ±    0.338  ns/op
-ArrayBufferBenchmark.addAll               100  avgt   30     266.845 ±    0.661  ns/op
+ArrayBufferBenchmark.addAll               100  avgt   30     265.249 ±    0.397  ns/op
-ArrayBufferBenchmark.addAll              1000  avgt   30    1690.149 ±    3.764  ns/op
+ArrayBufferBenchmark.addAll              1000  avgt   30    1712.475 ±    2.376  ns/op
-ArrayBufferBenchmark.addAll             10000  avgt   30   24878.508 ±   43.111  ns/op
+ArrayBufferBenchmark.addAll             10000  avgt   30   24931.691 ±   48.856  ns/op
-ArrayBufferBenchmark.filterInPlace         10  avgt   30      61.465 ±    0.099  ns/op
+ArrayBufferBenchmark.filterInPlace         10  avgt   30      61.529 ±    0.166  ns/op
-ArrayBufferBenchmark.filterInPlace        100  avgt   30     365.526 ±    0.842  ns/op
+ArrayBufferBenchmark.filterInPlace        100  avgt   30     377.592 ±    0.657  ns/op
-ArrayBufferBenchmark.filterInPlace       1000  avgt   30    3248.895 ±    5.395  ns/op
+ArrayBufferBenchmark.filterInPlace       1000  avgt   30    3054.808 ±   40.811  ns/op
-ArrayBufferBenchmark.filterInPlace      10000  avgt   30   37520.257 ±   62.193  ns/op
+ArrayBufferBenchmark.filterInPlace      10000  avgt   30   36584.331 ±   98.728  ns/op
-ArrayBufferBenchmark.flatMapInPlace1       10  avgt   30     292.213 ±    1.564  ns/op
+ArrayBufferBenchmark.flatMapInPlace1       10  avgt   30     293.366 ±    4.643  ns/op
-ArrayBufferBenchmark.flatMapInPlace1      100  avgt   30    2144.930 ±   17.724  ns/op
+ArrayBufferBenchmark.flatMapInPlace1      100  avgt   30    2194.724 ±    2.738  ns/op
-ArrayBufferBenchmark.flatMapInPlace1     1000  avgt   30   20215.275 ±  106.838  ns/op
+ArrayBufferBenchmark.flatMapInPlace1     1000  avgt   30   20318.600 ±   70.100  ns/op
-ArrayBufferBenchmark.flatMapInPlace1    10000  avgt   30  210170.468 ± 1116.711  ns/op
+ArrayBufferBenchmark.flatMapInPlace1    10000  avgt   30  210699.714 ±  389.451  ns/op
-ArrayBufferBenchmark.iteratorA             10  avgt   30      46.418 ±    0.087  ns/op
+ArrayBufferBenchmark.iteratorA             10  avgt   30      51.599 ±    0.116  ns/op
-ArrayBufferBenchmark.iteratorA            100  avgt   30     185.297 ±    0.470  ns/op
+ArrayBufferBenchmark.iteratorA            100  avgt   30     229.769 ±    0.669  ns/op
-ArrayBufferBenchmark.iteratorA           1000  avgt   30    1449.335 ±    3.177  ns/op
+ArrayBufferBenchmark.iteratorA           1000  avgt   30    1977.957 ±    4.647  ns/op
-ArrayBufferBenchmark.iteratorA          10000  avgt   30   17397.759 ±  151.373  ns/op
+ArrayBufferBenchmark.iteratorA          10000  avgt   30   22770.041 ±   37.288  ns/op
-ArrayBufferBenchmark.iteratorB             10  avgt   30      82.953 ±    0.259  ns/op
+ArrayBufferBenchmark.iteratorB             10  avgt   30      93.475 ±    0.242  ns/op
-ArrayBufferBenchmark.iteratorB            100  avgt   30     537.971 ±   10.782  ns/op
+ArrayBufferBenchmark.iteratorB            100  avgt   30     583.337 ±    1.681  ns/op
-ArrayBufferBenchmark.iteratorB           1000  avgt   30    4886.867 ±   31.022  ns/op
+ArrayBufferBenchmark.iteratorB           1000  avgt   30    5319.587 ±   96.622  ns/op
-ArrayBufferBenchmark.iteratorB          10000  avgt   30   48486.979 ±  153.746  ns/op
+ArrayBufferBenchmark.iteratorB          10000  avgt   30   52150.922 ±   75.137  ns/op
-ArrayBufferBenchmark.reverseIteratorA      10  avgt   30      44.654 ±    0.087  ns/op
+ArrayBufferBenchmark.reverseIteratorA      10  avgt   30      48.189 ±    0.166  ns/op
-ArrayBufferBenchmark.reverseIteratorA     100  avgt   30     170.628 ±    0.489  ns/op
+ArrayBufferBenchmark.reverseIteratorA     100  avgt   30     165.333 ±    0.456  ns/op
-ArrayBufferBenchmark.reverseIteratorA    1000  avgt   30    1332.711 ±    2.200  ns/op
+ArrayBufferBenchmark.reverseIteratorA    1000  avgt   30    1158.948 ±   11.021  ns/op
-ArrayBufferBenchmark.reverseIteratorA   10000  avgt   30   18947.001 ±  132.302  ns/op
+ArrayBufferBenchmark.reverseIteratorA   10000  avgt   30   16477.355 ±   86.881  ns/op
-ArrayBufferBenchmark.reverseIteratorB      10  avgt   30      89.263 ±    0.214  ns/op
+ArrayBufferBenchmark.reverseIteratorB      10  avgt   30      97.480 ±    0.247  ns/op
-ArrayBufferBenchmark.reverseIteratorB     100  avgt   30     504.658 ±   30.568  ns/op
+ArrayBufferBenchmark.reverseIteratorB     100  avgt   30     618.153 ±    9.257  ns/op
-ArrayBufferBenchmark.reverseIteratorB    1000  avgt   30    4033.524 ±  267.711  ns/op
+ArrayBufferBenchmark.reverseIteratorB    1000  avgt   30    5465.261 ±    7.082  ns/op
-ArrayBufferBenchmark.reverseIteratorB   10000  avgt   30   43521.916 ± 1915.392  ns/op
+ArrayBufferBenchmark.reverseIteratorB   10000  avgt   30   50848.591 ±  109.594  ns/op
-ArrayBufferBenchmark.update                10  avgt   30      41.986 ±    0.110  ns/op
+ArrayBufferBenchmark.update                10  avgt   30      42.774 ±    0.162  ns/op
-ArrayBufferBenchmark.update               100  avgt   30     192.961 ±    0.586  ns/op
+ArrayBufferBenchmark.update               100  avgt   30     209.066 ±    0.539  ns/op
-ArrayBufferBenchmark.update              1000  avgt   30    1592.343 ±    2.882  ns/op
+ArrayBufferBenchmark.update              1000  avgt   30    1736.486 ±    7.763  ns/op
-ArrayBufferBenchmark.update             10000  avgt   30   17729.414 ±   33.149  ns/op
+ArrayBufferBenchmark.update             10000  avgt   30   19053.384 ±   35.325  ns/op
```

NthPortal commented:
my interpretation of the benchmarks is that there is a small penalty to iterators, a very small, probably negligible effect on single-change methods (`update`) that's only noticeable at all when done at huge scale, and no meaningful effect on bulk-change methods. I'm personally okay with the small penalty to iterators in exchange for better safety.

I still don't understand the `reverseIteratorA` results

Ichoran commented:
The desktop Java 16 numbers look more like I'd expect.  Seems okay.

For the record, I do not think a ~5-10% speed penalty for traversal/indexing-heavy operations on a collection whose main purpose is to speed up traversal/indexing-heavy operations, is a good tradeoff for the safety, mostly because I don't think the safety issue is remotely common.  But reasonable people can disagree on this matter, and I do agree that the penalty is fairly small.

NthPortal commented:
there seemed to have been less of a cost with `ListBuffer` than here. any ideas about what we could do differently to have less of a performance cost? should we make the iterator more bespoke, so it has direct access to the field?

Ichoran commented:
It seems a reasonable thing to try, but I have low confidence in any particular fix.  Fairly often, improvements that seem like a clear win end up making the JIT compiler take a different optimization strategy, which kind of randomizes the performance.  Not much to do but try and see.

I certainly _have_ sped code up before by having direct monomorphic access to fields.  That was one key trick in turning in good performance on some tasks in the Computer Languages Benchmark Game.  But in my own code the results have been more mixed--yes, sometimes it works, sometimes not.  Sometimes not.  And I never feel motivated to dig into the details to the incredible depth that Aleksey Shipilëv does.  I usually tell myself some story about "method now has more bytecode" to make myself feel better.

NthPortal commented:
that change did not seem to be more performant (though I forgot to stop a service running on the server first, so who knows :woman_shrugging:). regardless, it did not end up being binary compatible, so it's a moot point.

I'd be interested in getting someone else's thoughts on the performance costs. I'm obviously biased in that I wrote the code. I'm basically copying what Java did (or trying my best to), but I've no idea where to look to find the performance costws Java had from it, if they even published them.

SethTisue commented:
I don't have a _strong_ opinion, but I'm inclined to go ahead and merge at this point (I mean, after allowing until Friday for feedback, say), as I feel we've already reached roughly the same point as #9174 was at when we went ahead and merged it.

SethTisue commented:
@scala/collections anyone see anything here they want to dig deeper into?

SethTisue committed:
Merge branch '2.13.x' into topic/safe-iterators-2/PR

joroKr21 commented:
Sorry, missclick on my phone - I'm commuting hence bored and browsing GH notifications :smile:

joroKr21 commented:
TIL that you can disable it from the phone but not set it back
